### DISCUSSION TITLE
Parallel db queries

### DISCUSSION THREAD

**Discussion by jsbrain** [directus/directus#16018] [Archive]
If I'm not mistaken, queries in e.g. createMany [are handled sequentially](https://github.com/directus/directus/blob/73fa8d66acb1e05f015adbb1598229afae69ca08/api/src/services/items.ts#L272) while they probably could be handled in parallel via `Promise.all()` which would drastically improve performance on large queries.

Is there any particular reason for that?

**Comment by rijkvanzanten** [comment]
Parallel queries can cause infinite hangs on transactions when a single db connection is used (as is the case with DB_POOL_MAX=1 or SQLite usage)

**Reply by jsbrain** [reply]
I see. But shouldn't there be a option to change this then for users that don't have this restriction?

Writing 200 records to the db takes around 6 seconds while it should take less than 100ms.

I have much larger datasets that need to be updated periodically so this is only getting worse from here.

So far my only option is to write straight to the db which I feel is a bit of a shame.

**Reply by jsbrain** [reply]
Sounds like simply checking wether DB_POOL_MAX=1 or sqlite to enable/disable parallel execution accordingly would do the trick?

**Reply by rijkvanzanten** [reply]
"simply" being the keyword there ha

**Reply by jsbrain** [reply]
Based on my initial question and your statement implicating that the only reason not to enable parallel execution is that there are edge cases, where parallel execution can cause infinite hangs and performing a check wether the environment would allow parallel execution can be achieved with a one-liner aka

```ts
const enableParallel = config.db !== 'sqlite' && config.DB_POOL_MAX > 1;
```

I indeed would consider this simple.

The actual implementation of the parallel execution logic is probably a different story but that's not the topic of this discussion as of now.

In case there is more to it which you didn't tell me I'm eager to hear about I'll gladly reconsider my statement.

**Reply by rijkvanzanten** [reply]
> The actual implementation of the parallel execution logic is probably a different story but that's not the topic of this discussion as of now.

Ehh yeah that's where it becomes _not simple_! 😄 Of course your "should it be enabled?" line is simple, but then using that logic and passing it around to all the places where queries are made, and then batching the actual queries according to that logic is the complicated bit. That's very much related to this discussion IMO 🙂

**Reply by jsbrain** [reply]
I asked if there is a reason why parallel execution is prevented. You answered that the reason is that there are some edge cases where this would cause problems. In response, I said that one could detect these edge cases by _simply_ checking whether they apply or not, so that there could be a decision path whether to allow parallel execution or not.

I asked about WHY, not HOW. Of course, the final implementation may very well have something to do with the problem, but that was not the point of my statement, so I don't understand why you are trying to ridicule it.

I apologize if I offended you by using _"simply"_ in my statement. That was not my intention.
